### ISSUE TITLE
Google auth broken for new apps?

### ISSUE THREAD

**Issue by toddmedema** [EddyVerbruggen/cordova-plugin-googleplus#311]
We had auth working on the iOS build of our app (http://app.expeditiongame.com), and at some point it stopped working, so I've been trying to revive it.

It's currently getting a network error 400 `"Invalid parameter value for origin: Missing authority: file://"`when checking with Google if the user is signed in, `https://accounts.google.com/o/oauth2/iframerpc?action=checkOrigin&origin=file%3A%2F%2F&client_id=545484140970-r95j0rmo8q1mefo0pko6l3v6p4s771ul.apps.googleusercontent.com`

So, if anyone has insight into that bug (been Googling it for an hour with no luck), it would be appreciated. I tried every solution I saw, but none worked...

And then, I came across an announcement from the Google Auth team that **In the coming months, we will no longer allow OAuth requests to Google in embedded browsers known as “web-views”** - https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html

Which makes me think that maybe I'm not crazy, and that Google pulled the carpet out from under me / removed support for it? Are there any plans for this plugin to support the new method of auth, or what would be a good new plugin to use? OR, am I crazy, this plugin already supports the new way of doing things, and I just need to figure out the bug mentioned above?

Thank you for your help!

**Comment by sammuggPS** [comment]
I saw that article when it came out, but didn't remember it until now. Thanks for posting about it! @EddyVerbruggen, it looks like this is something that will have to be reviewed. It seems that hybrid login with cordova isn't going to work moving forward.

**Comment by EddyVerbruggen** [comment]
AFAIK Google Sign In is (which we use) is exactly the solution for this problem; the in-app browser is a no-go, but leveraging Safari(ViewController) on iOS (which Google Sign In uses) is supposed to be the way forward.

@toddmedema Which version of this plugin do you use?

**Comment by sammuggPS** [comment]
Good catch, @EddyVerbruggen. I missed that on my quick pass over the post. I believe this compromises backwards compatibility for iOS < v9.0

**Comment by toddmedema** [comment]
@EddyVerbruggen I have 5.1.1 installed

**Comment by toddmedema** [comment]
Are other people having issues with iOS too, or am I the odd one out? Our iOS app is currently broken, so would appreciate any help or debugging tips you'd be able to provide :)

**Comment by jpduckwo** [comment]
I'm about to try out the plugin on iOS and android in a new installation. Have you had any luck yet @toddmedema? I will let you know how I go.

**Comment by toddmedema** [comment]
I haven't yet - had plenty of other bugs to work on in the mean time :) Let me know if you have luck with it! Our auth's working on Android, just not iOS

**Comment by jpduckwo** [comment]
@toddmedema I got it working on iOS. One thing to check is the enabled capabilities in your iOS app. The plugin tries to enable the app for Keychain Access which might not work if you haven't setup your provisioning profile for it. Anyway double check this setting is enabled in your Xcode  project capabilities.

![image](https://cloud.githubusercontent.com/assets/13150911/20998663/90b8b41c-bd63-11e6-9e89-a40446c836f3.png)

**Comment by valgen** [comment]
@jpduckwo 
I'm working on a Visual Studio 2015 Cordova project connected to a Mac iOS 10 and iPhone 6 Simulator (remotebuild).
How to access the Xcode project capabilities?

**Comment by jpduckwo** [comment]
Took me a while to find too...

Check this article: https://developer.apple.com/library/content/documentation/IDEs/Conceptual/AppDistributionGuide/AddingCapabilities/AddingCapabilities.html

**Comment by toddmedema** [comment]
Hmm. Still no luck, and keychain sharing is enabled. Maybe I'm barking up the wrong tree, but been experimenting with keys and client ID's, just been trying to figure out what's different about the iOS client vs the Android and web clients...

Looking at the Google Developers console (console.developers.google.com/apis/credentials), there are API keys and OAuth 2.0 client IDs. I've tried several combinations of them, but I always get the same auth error on my server, from inside the `passport-google-id-token` module: `JsonWebTokenError: jwt audience invalid. expected: 54*********ul.apps.googleusercontent.com` (client ID redacted)

I get that error even when that I build the iOS app with that client ID - whether I'm building it with our server API key or the auto-generated iOS API key.

What are other people doing for their keys? 

Any other things I can try?

Also, just to make sure: it works for you in your emulator, right? (I don't currently have a physical iOS device to test on)

**Comment by toddmedema** [comment]
Update: this is strange. If I change the OAuth client ID on my server to use the iOS client ID instead of the server client ID, it works! Of course, then web and Android break.

Which, now that I think about it, makes sense. The cordova config for this plugin includes the iOS-specific `REVERSED_CLIENT_ID` that's the iOS client ID.

So, I tried changing the REVERSE_CLIENT_ID to use the server ID, and totally rebuilding the iOS app... but then I get the error (when trying to log in to the app) `Error: redirect uri mismatch`, that the redirect uri doesn't match the ones authorized for the OAuth client.

So, looks like REVERSE_CLIENT_ID needs to be set to the iOS client ID. But, in doing so, it also means my server needs to also use the iOS client ID? But, also use the server ID, so that web and Android apps can log in?

Needless to say, pretty confused and would appreciate any guidance :)

**Comment by toddmedema** [comment]
Any ideas? Even just on other things I could try / test / log? Still stumped on this one

**Comment by jpduckwo** [comment]
Hi @toddmedema, 

One mistake I made was using the wrong client id. So in the config.xml you need to set the iOS client id, but if you want to get an authcode back, you need to pass a Web Client ID with the function call 
```
                $window.plugins.googleplus.login({
                        // 'scopes': 'https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/userinfo.profile',
                        'webClientId': 'xxxx-different-to-ios-one-xxxx.apps.googleusercontent.com',
                        // 'offline': false,
                    },
                    function(obj) {
                        $log.debug('googleAuth: Cordova method object returned: ', obj);
                        var credentials = {
                            // obj.email
                            // obj.userId
                            // obj.displayName
                            // obj.familyName
                            // obj.givenName
                            // obj.imageUrl
                            // obj.idToken
                            // obj.serverAuthCode
                            primaryEmail: obj.email,
                            authToken: obj.idToken,
                            firstName: obj.givenName,
                            lastName: obj.familyName,
                        };
                        return deferred.resolve(credentials);
                    },
                    function(msg) {
                        return deferred.reject(msg);
                    }
```
This is generated or copied from https://console.developers.google.com/apis/credentials. In the list you should see your iOS client id too... don't use that in the code, only in the config.xml.

**Comment by toddmedema** [comment]
When I search for the iOS key in the main project folder, I find it in the following files:

```
config.xml
plugins/ios.json
platforms/ios/ios.json
platforms/ios/Expedition/Expedition-Info.plist
platforms/ios/build/emulator/Expedition.app/Info.plist
```

I only defined it in config.xml as the REVERSED_CLIENT_ID, the others look to be entirely auto-populated. 

The login function then uses the web client id as suggested:
```
function loginCordova(cb: (user:UserState)=>any) {
  var that = this;
  window.plugins.googleplus.login({
    scopes: authSettings.scopes,
    webClientId: authSettings.clientId,
  }, function(obj: any) {
    registerUserAndIdToken({
      name: obj.displayName,
      image: obj.imageUrl
    }, obj.idToken, cb);
  }, function(msg: string) {
    throw new Error(msg);
  });
}
```

In digging in further, I think the problem might be because we have an auth server, as well as auth in the app. Looks like the app is actually authing just fine with iOS!

Once our users have authenticated with Google, they send that auth info to our server, so that they have permission to hit our API. Which works fine with the web app and Android b/c they're using the same API key as the server. But, because the iOS app is using the REVERSED_CLIENT_ID, it doesn't validate on the server, b/c the server is using the web app key. So, seems like we have to set up the server to auth against both the web and iOS keys?

**Comment by jpduckwo** [comment]
I don't think the client ID is used on the server side is it? All you need to do is verify the auth token on the server side. https://developers.google.com/identity/sign-in/android/backend-auth see the heading 'Verify the integrity of the ID token'

**Comment by antpuleo2586** [comment]
The ionic1 (and ionic2) implementation of Native Google sign in is on an ionic docs page:

https://docs.ionic.io/services/auth/google-native.html

They suggest switching from their InAppBrowser implementation (https://docs.ionic.io/services/auth/google-auth.html) because of the support dropping on 20th April 2017.

**Comment by toddmedema** [comment]
@antpuleo2586 yep - @EddyVerbruggen talked about that earlier in this thread, it sounds like they're already doing it the correct / new way?

**Comment by toddmedema** [comment]
Fixed! Changed the server to try validating against both the web and iOS keys
